Validate collection names

seeforschauer opened the pull request:
# Description

This PR closes #806.

w84thesun commented on internal/handlers/pg/pgdb/pool.go:
```diff
@@ -343,10 +347,38 @@ func (pgPool *Pool) DropDatabase(ctx context.Context, db string) error {
+// Collection name must not contain non-latin letters, spaces, dots, dollars, dashes,
+// name collection name length must be less than 119 characters.
+func (pgPool *Pool) validateCollectionName(collection string) error {
+	if len(collection) == 0 {
+		return ErrInvalidTableName
+	}
+
+	if len(collection) > 119 {
+		return ErrInvalidTableName
+	}
```
I think those are redundant checks, as they are included in regex. Therefore they can be removed.

w84thesun reviewed (changes requested):
LGTM, except collection length checks.

AlekSi commented on integration/basic_test.go:
```diff
@@ -128,22 +128,74 @@ func TestFindCommentQuery(t *testing.T) {
+			"TooLongForBoth": {
+				collection: "very_long_collection_name_that_fails_both_in_mongo_and_in_ferretdb_databases" +
+					"_for_ferretdb_it_fails_because_it_is_more_than_119_characters__for_mongo_it_fails_because_it_is_more_than_255_charachters_" +
+					"long_that_excludes_non_latin_letters_spaces_dots_dollars_dashes",
```
Unless we really want people to actually read it, it could be replaced with `const collectionName300 = strings.Repeat("a", 300)` and referring to it both there and below, in the error message

The same in other places

seeforschauer replied:
> Unless we really want people to actually read it, it could be replaced with `const collectionName300 = strings.Repeat("a", 300)` and referring to it both there and below, in the error message
> 
> The same in other places

Changed here and in other places for collection name tests.

AlekSi commented on integration/basic_test.go:
```diff
@@ -128,22 +128,74 @@ func TestFindCommentQuery(t *testing.T) {
+					Code:    73,
+					Message: `Invalid collection name: collection_name_with_a-$`,
+				},
+				alt: "Collection must not contain non-latin letters, spaces, dots, dollars, dashes and be longer than 119 characters.",
```
Can we return the same error message there and in other places?

seeforschauer replied:
> Can we return the same error message there and in other places?

there -> where?
other places - > which places do you mean?

AlekSi replied:
As we discussed previously, we should return the same error message as MongoDB where we can. I this test case MongoDB returns `Invalid collection name: <collection name>`. That's a good correct error message. It should not be hard to make FerretDB return the same error message. So let's do that here and in other places in this PR.

AlekSi replied:
Let's keep sentinel error `ErrInvalidTableName` as is for now then, and not add a new type

AlekSi commented on integration/basic_test.go:
```diff
@@ -128,22 +128,74 @@ func TestFindCommentQuery(t *testing.T) {
+			name, tc := name, tc
+			t.Run(name, func(t *testing.T) {
+				err := collection.Database().CreateCollection(ctx, tc.collection)
+				assert.NotNil(t, tc.err)
```
Not required with the helper below

seeforschauer replied:
> Not required with the helper below

Sorry I didn't understand you. WDYM?

I meant by this line the following:
This:
```go
assert.NotNil(t, tc.err)
```
Means checking that we didn't forget to add an error in the test cases for errors.

AlekSi replied:
That code does not stop the execution of the test. If `tc.err` is nil, then `*tc.err` below will panic

seeforschauer replied:
> > That code does not stop the execution of the test. If `tc.err` is nil, then `*tc.err` below will panic
> 
> Thank you. Changed to `require`.

removed it all in all.

AlekSi commented on internal/handlers/pg/pgdb/pool.go:
```diff
@@ -343,10 +347,38 @@ func (pgPool *Pool) DropDatabase(ctx context.Context, db string) error {
+// validateCollectionName validates collection name.
+// Collection name must not contain non-latin letters, spaces, dots, dollars, dashes,
+// name collection name length must be less than 119 characters.
+func (pgPool *Pool) validateCollectionName(collection string) error {
```
Let's inline this method into the only place where we use it

seeforschauer replied:
> Let's inline this method into the only place where we use it

removed.

AlekSi commented on internal/handlers/pg/pgdb/pool.go:
```diff
@@ -343,10 +347,38 @@ func (pgPool *Pool) DropDatabase(ctx context.Context, db string) error {
+		return ErrInvalidTableName
+	}
+
+	regex := regexp.MustCompile("^[a-zA-Z_][a-zA-Z0-9_]{0,119}$")
```
We could compile it only once

seeforschauer replied:
> We could compile it only once

moved to package level private variables.

AlekSi commented on internal/handlers/pg/pgdb/pool.go:
```diff
@@ -343,10 +347,38 @@ func (pgPool *Pool) DropDatabase(ctx context.Context, db string) error {
+
 // CreateCollection creates a new FerretDB collection in existing schema.
 //
 // It returns ErrAlreadyExist if table already exist, ErrTableNotExist is schema does not exist.
```
It can return other errors too now, we should document that

seeforschauer replied:
> It can return other errors too now, we should document that

added to doc and checked how the doc looks like.

rumyantseva commented on README.md:
```diff
@@ -33,6 +33,12 @@ and [contributing guidelines](CONTRIBUTING.md).
+* collection name cannot start with the reserved prefix `_ferretdb_`.
+* collection name must not include non-latin letters, spaces, dots, dollars or dashes.
+* collection name length must be less than 119 symbols.
+
```
There is one more condition - collection name must not start with a number.

rumyantseva commented on internal/handlers/pg/pgdb/pool.go:
```diff
@@ -345,8 +352,19 @@ func (pgPool *Pool) DropDatabase(ctx context.Context, db string) error {
+		return ErrInvalidTableName
+	}
+
+	if strings.HasPrefix(collection, collectionPrefix) {
```
With that, I wonder if the `collectionPrefix` constant should be called something like `reservedCollectionPrefix`. But I don't insist :)

seeforschauer commented:
@rumyantseva 

> 1. In the PR body there is a case `max length in FerretDB: 255, in MongoDB: 63.`.

Please pay attention to the caption - it's a difference between FerretDB and MongoDB. 
I listed these differences in the PR body to remember what cases to cover, document, and mention in the dance tests. 

>  Should it be `max length in FerretDB: 119`? or what 119 is about? :)

Not sure about 119, it's a magic number for now. And yes, this magic number is not in the constant somewhere, I follow the task guideline that we can change it later.  

> 2. Are dance cases in the scope of this task? Or is there a separate issue for it?

It is in the scope of the task and it is not in the scope of the PR since the repository is FerretDB. 

> 3. It's great that there is a new piece of doc for README. It's not 100% full/accurate (I added a couple of comments about it).

Added, thank you for noting this.

> 4. Special thanks for updating `CreateCollection` godoc.

You are welcome ^ - ^. I'm glad I have a command to check what docs look like (thanks to @AlekSi).

seeforschauer commented on README.md:
```diff
@@ -33,6 +33,12 @@ and [contributing guidelines](CONTRIBUTING.md).
+
+* collection name cannot start with the reserved prefix `_ferretdb_`.
+* collection name must not include non-latin letters, spaces, dots, dollars or dashes.
+* collection name length must be less than 119 symbols.
```
> "Less" or "less or equal"? The regex is about "less or equal".

changed.

rumyantseva reviewed (approved):
LGTM

AlekSi commented on internal/handlers/pg/pgdb/pool.go:
```diff
@@ -345,8 +352,19 @@ func (pgPool *Pool) DropDatabase(ctx context.Context, db string) error {
 //
-// It returns ErrAlreadyExist if table already exist, ErrTableNotExist is schema does not exist.
+// It returns:
+//  - ErrInvalidTableName - if the table name doesn't conform to restrictions.
```
https://github.com/golang/proposal/blob/master/design/51082-godocfmt.md#new-syntax-for-lists

> To satisfy both, all three bullets and both forms of numbers will be recognized, but gofmt (see below) will rewrite them to a single canonical form: dash for bullets, and “N.” for numbers.

In short, it does not matter now, but the next version of Go will reformat all comments to use dashes

AlekSi commented on README.md:
```diff
@@ -33,6 +33,13 @@ and [contributing guidelines](CONTRIBUTING.md).
+
+* collection name cannot start with the reserved prefix `_ferretdb_`.
+* collection name must not include non-latin letters, spaces, dots, dollars or dashes.
+* collection name length must be less or equal than 119 symbols.
```
120, actually. The length should be in the range [1-120] inclusively

seeforschauer replied:
> 120, actually. The length should be in the range [1-120] inclusively

gotcha. I thought what a strange number 119. fixed.

AlekSi commented on integration/basic_test.go:
```diff
@@ -128,22 +129,73 @@ func TestFindCommentQuery(t *testing.T) {
+	t.Run("Ok", func(t *testing.T) {
+		ctx, collection := Setup(t)
+
+		longCollectionName := strings.Repeat("a", 65)
+		err := collection.Database().CreateCollection(ctx, longCollectionName)
+		require.NoError(t, err)
+		sixtyThreeCharsCollectionName := strings.Repeat("a", 63)
+		err = collection.Database().CreateCollection(ctx, sixtyThreeCharsCollectionName)
+		require.NoError(t, err)
```
65 and 63 look special, but it is not clear why, and both work. Do we need both? Could we use a simpler number like 100 that is also not special?

seeforschauer replied:
> 65 and 63 look special, but it is not clear why, and both work. Do we need both? Could we use a simpler number like 100 that is also not special?

I've left one only.
